### PR TITLE
Request for feedback - Text expansion/shortcuts in text areas

### LINKED ISSUE #130: Exam and procedural notes shortcuts
# APR = Standard notes for APR

The notion behind this is to create a shortcut process for copy where using the hashtag symbol will auto fill-in a pre-configured note. So to accomplish this feature, we'd want to do at least these things:
- [ ] Create a text area component that is "aware" of shortcuts.
- [ ] Create an administrative screen for creating said shortcuts.
- [ ] retrofit at least the Patient, Visit, and Appointment fields to use the shortcut-aware component.

### PR DESCRIPTION
**Pull request by eddiesholl**
Fixes #130

**Changes proposed in this pull request:**
- New 'miscellaneous' admin page, for creating shortcuts
- New component 'expand-text' component, which wraps an em-text
- Attach a floating div below the em-text with user feedback for possible expansions
- Expansions activated by typing '#foo' and then pressing enter

At this stage I'm looking for any feedback on my proposed solution. It's my first pull request here and a non-trivial feature. There's still some obvious outstanding items, including:
- A bit more i18n required
- Floating div does not resize with the text area
- Clean up text parsing methods
- Extract the inline style work in didInsertElement
- More tests

cc @HospitalRun/core-maintainers

### CODE REVIEW

**Comment by tangollama** [comment]
@eddiesholl thanks for submitting this. I'll try to look through this tonight or tomorrow and get you back some feedback.

**Comment by tangollama** [comment]
@eddiesholl, apologies for not getting to this over the weekend. I just reviewed the functionality. From a functional perspective, this is right on! There are two steps I want to confirm and then one issue I uncovered. 

First, I need to confirm with @jkleinsc that the component meets with his blessing. We should be able to handle that tomorrow.

Second, we need to run the UX issues through @jglovier. To that end, we need to include screenshots from the admin screen as well as usage in the procedures screen where this branch implements the expand-text component. I've done that here.

screen shot 2016-05-02 at 11 55 17 pm
screen shot 2016-05-02 at 11 55 38 pm

@jglovier, if you have feedback / concerns, can you respond please?

Finally, the issue I uncovered was how the type-ahead recommendation presents itself when the input is in a modal. I implemented the component in the patients/notes template (which presents as a modal) and it positions itself "behind" the modal. So that's a limitation. Do you think you can address this or do we need some secondary CSS help?
screen shot 2016-05-02 at 11 56 05 pm

**Comment by eddiesholl** [comment]
Cheers for the feedback, yes there's still some cleanup to do, and hence the request for feedback, some input required on things like the styling and UI.

The main decision to make is on the way feedback is presented. I went with floating a div below the textarea for a few reasons:

* Trying to maintain support for dropping into ember rapid forms
* A form of feedback that is generally not visible, and not too distracting when it is relevant
* No dynamic resizing or jumping of the page

A fancier solution is to actually place the feedback over the text, more like an intellisense solution, like 'multiple selection' at https://leaverou.github.io/awesomplete/#advanced-examples. But that's more work and more fiddly, and requires some hackery over a textarea.

The wierdness in the modal dialog is obviously going to need some changes. There's also an outstanding issue that resizing the textarea doesnt resize the feedback area.

Secondly, the admin area was all a bit improvised. I didn't want to create an entire admin page just for it, and hence 'miscellaneous', but to be honest there's enough controls there to imply its own page.

**Comment by jkleinsc** [comment]
What I see so far looks good, though the UI does need some clean up.

Also, I would make the Text replacements its own admin page instead of calling it Miscellaneous.  At some point we probably need to better arrange admin pages, but for now I would just keep it as an additional page called Text Replacements instead of Miscellaneous.

File: app/admin/textrepl/controller.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import Ember from 'ember';
+import EmberValidations from 'ember-validations';
+
+export default Ember.Controller.extend(EmberValidations, {
+  hideCancelButton: true,
+  updateCapability: 'update_config',
+
+  createExpansion: function() {
```
Comment (billybonks) [inline app/admin/textrepl/controller.js]: we should be using `on(function(){})` format as the current syntax will eventually be removed. you can export the on as well by using `const { on } = Ember`

File: app/admin/textrepl/controller.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    const newExpansion = this.get('store').createRecord('text-expansion');
+    this.set('newExpansion', newExpansion);
+  }.on('init'),
+
+  validations: {
```
Comment (billybonks) [inline app/admin/textrepl/controller.js]: generally i add validations at the end @jkleinsc there was discussion of adding a style guide

File: app/admin/textrepl/route.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import AbstractIndexRoute from 'hospitalrun/routes/abstract-index-route';
+
+import { translationMacro as t } from 'ember-i18n';
+export default AbstractIndexRoute.extend({
+  pageTitle: t('admin.textrepl.page_title'),
+  hideNewButton: true,
+
+  model: function() {
+    const store = this.get('store');
+    store.findAll('text-expansion');
```
Comment (billybonks) [inline app/admin/textrepl/route.js]: why do you call this twice ?

File: app/admin/textrepl/template.hbs [hunk 4]
```diff
@@ -0,0 +1,35 @@
+<div class="panel panel-primary">
+<h3>{{t 'admin.textrepl.existing_repl'}}</h3>
+<p>{{t 'admin.textrepl.repl_desc'}}</p>
+<table class="table">
+  <tr class="table-header">
+    <th style="min-width: 150px">{{t 'labels.from'}}</th>
```
Comment (billybonks) [inline app/admin/textrepl/template.hbs]: inline styles 😭 @jglovier

File: app/locales/en/translations.js [hunk 5]
```diff
@@ -35,7 +35,8 @@ export default {
       load_db: 'Load DB',
       users: 'Users',
       new_user: 'New User',
-      user_roles: 'User Roles'
+      user_roles: 'User Roles',
+      textrepl: 'Text Replacements'
```
Comment (billybonks) [inline app/locales/en/translations.js]: `text_replacement`

File: app/admin/textrepl/template.hbs (hunk 4 above)
Comment (jglovier) [inline app/admin/textrepl/template.hbs]: Yes, please no inline styles. Add a class and then style the class in an existing or new SCSS file please. :bow:

File: app/styles/components/_form_styles.scss [hunk 6]
```diff
@@ -6,3 +6,13 @@
 
   select { height: 44px; }
 }
+
+.text-expansion-div {
```
Comment (jglovier) [inline app/styles/components/_form_styles.scss]: To keep classnames clean, lets change this to just `.text-expansion` or `text-expansion-container`.

**Comment by jglovier** [comment]
>if you have feedback / concerns, can you respond please?

In general, I think the flow here is pretty good. The snippet dialog needs style help, and I'll mock up something shortly as a guide.

Also, it might be helpful to add a shortcut to "Add new expansion" inside the dialog (see the GitHub comment bar saved replies dialog for an example).

image


>it positions itself "behind" the modal.

Would setting a high `z-index` value remedy this, or is the markup nested to deep to solve that way?

**Comment by jglovier** [comment]
Quick pass at styling for the expansion dialog.

image

I could go either way on the "add new expansion" piece, but if we punt on that now I think it's worth revisiting later.

**Comment by tangollama** [comment]
@eddiesholl is there any additional feedback you're looking for at the moment? Just checking in.

**Comment by tangollama** [comment]
@eddiesholl I was trying to review this branch and there are some merge conflicts with master that *seem* to be keeping me from reviewing the branch now. Can you update from master on your branch so that I can see the feature in it's current state?

Sidenote: we're getting a lot of positive feedback from clinicians on this work. I think people are really going to value this.

**Comment by tangollama** [comment]
This code has been incorporated into the master via an alternative branch. Closing the PR.

**Pull request closed without merging**